specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,174 @@
 
 ## Overview
 
-// TODO: this is a stub.
+Hooks are a mechanism whereby application developers can add arbitrary behavior
+to flag evaluation. They operate similarly to middleware in many web frameworks.
+
+### Definitions
+
+**Hook**: Application author/integrator-supplied logic that is called by the OpenFeature framework at a specific stage.
+**Stage**: An explicit portion of the flag evaluation lifecycle. e.g. `before` being "before the evaluation is run.
+**Invocation**: A single call to evaluate a flag. `client.getBooleanValue(..)` is an invocation.
+**API**: The global API singleton.
+
+### Hook context
+
+Hook context exists to provide hooks with information about the invocation.
+
+###### Requirement 1.1
+
+> Hook context **MUST** provide: the flag key, evaluation context, default value and a list of  executed hooks by stage.
+
+
+##### Condition 1.2
+
+> If the language type system differentiates between strings, numbers, booleans, and structures.
+>
+>##### Condition 1.2.1
+>
+> Condition: You **MUST** provide `flag type`.
+
+
+##### Requirement 1.2
+
+> Hook context **SHOULD** provide: provider, client
+
+
+##### Requirement 1.3
+
+> flag key, flag type, default value properties **MUST** be immutable. If the language does not support immutability, the hook **MUST NOT** modify these properties.
+
+##### Requirement 1.4
+
+> The evaluation context **MUST** be mutable only within the `before` hook.
+
+### HookHints
+
+##### Requirement 2.1
+
+> HookHints **MUST** be a map of objects.
+
+
+##### Condition 2.2
+
+> The implementation language supports a mechanism for marking data as immutable.
+>
+> ##### 2.2.1
+>
+> Condition: HookHints **MUST** be immutable.
+
+
+### Hook creation and parameters
+
+
+##### Requirement 3.1
+
+> Hooks **MUST** specify at least one stage.
+
+##### Requirement 3.2
+
+> The `before` stage **MUST** run before flag evaluation occurs. It accepts a `hook context` (required) and `state` (optional) as parameters and returns either a `HookContext` or nothing.
```

Comment:
Seems like we have a conflict here between `state` and `HookHints`? They are the same thing, right?